### PR TITLE
Fix small wall placement count preview (thanks to gynt)

### PR DESCRIPTION
**Pull request by LordHansCapon**
Fixes the bug where the small wall placement preview only allowed the player to use half of their stone.
See issue #663.

**Special thanks to gynt for finding and reverse-engineering the wall build preview!**

**Test build:**
https://we.tl/t-6SiSwVALlU

### CODE REVIEW

**Comment by LordHansCapon** [comment]
Looks like they touched this part of the code in SHC-E, as with 1 remaining stone, you can only place 3 small walls, and then 1 more small wall. Same change works flawlessly in SHC-HD though.

**Comment by LordHansCapon** [comment]
Fixed the SHC-E difference bug. Now it works in both HD and SHC-E.

**Test build:**
https://we.tl/t-6FrqCx8Hn0

**Comment by Krarilotus** [comment]
works as intended, tested also other walls to remain the same as before. tested with SHC and SHC-E

**Comment by GRhin** [comment]
This does not work in multiplayer, it lags the game, and only half the wall is placed.

File: UnofficialCrusaderPatch/Version.cs [hunk 1]
```diff
@@ -400,6 +400,44 @@ public class Version
...
             },
             
+            new Change("o_fix_small_wall_placement_count", ChangeType.Bugfix, true)
+            {
+                new DefaultHeader("o_fix_small_wall_placement_count")
+                {
+
+                    new BinaryEdit("o_fix_small_wall_placement_count_pre") // 437FB1
+                    {
+                        new BinAddress("CurrentlyControlledPlayer", 2),
```
Comment (LeSpec) [inline UnofficialCrusaderPatch/Version.cs:410]: CurrentlyControlledPlayer = 01A275DC
instead: 01A275D8 (player of the task being processed)

File: UnofficialCrusaderPatch/Version.cs [hunk 2]
```diff
@@ -400,6 +400,44 @@ public class Version
...
+                        new BinAddress("CurrentlySelectedWallID", 42),
```
Comment (LeSpec) [inline UnofficialCrusaderPatch/Version.cs:411]: instead of CurrentlySelectedWallID:
01959DF6 + 4f8 * [0194AF8C]

**Comment by LeSpec** [comment]
Note:
The game uses esi+constant to calculate those addresses. I don't know why, since esi seems to be always 0191D768.
E.g. esi+109E70 = 0191D768+109E70 = 01A275D8

**Comment by LeSpec** [comment]
I only checked which addresses should be used. Didn't check where or what the code is doing.

**Comment by gynt** [comment]
Nice work @LeSpec ! Thanks to that, we know what is going on here.

So, what is going on here is that the code checks _twice_ if building a wall is allowed, 1, at the UI level, and 2, at the game level. It calls the same function at some point for each of these levels.

The current fix grabs the player id and the current wall type from a function that is only called by the UI. This worked in single player, because the UI is always involved (currentUIPlayerID == actionPlayerID). In multiplayer, when your fellow player places something, your UI is not involved, so this trick does not work (currentUIPlayerID != actionPlayerID).

As LeSpec points out, we need the two variables from a different function. 
It is tricky, because what LeSpec suggests are addresses that are used after the action is already accepted by the UI, while this fix was specifically created to address the fact that the UI does not allow it at all (and the action is never created). 

The current fix is placed inside this function:
```cpp
getWallTilesThatCanBeBuilt(currentUIPlayerID) // in Single Player at the UI level, but not in Multi Player
getWallTilesThatCanBeBuilt(actionPlayerID) // in Single Player and Multi Player at the code level
```

As you can see the function is unaware of whether you are placing a low wall or a high wall, so the fix is to make the function aware of that fact by storing that value in memory and using it inside the function. The problem is, that value is in two different locations, depending on where we are at the UI level or the code level.

I think the most maintainable fix is to make this fix hook two locations, but then after `getWallTilesThatCanBeBuilt` is called, to adjust the outcome of the function in case of low wall placement. `getWallTilesThatCanBeBuilt` is called from 2 locations that are relevant (total of 3), 00437fd3 (ui level), 00502f85 (code level). The code level is easy to address, we should multiply by 2 if the action is a low wall command (that is in the stack, because it is a parameter to the function we are in). The ui level can keep its "CurrentlySelectedWallID" AOB.

**Comment by LeSpec** [comment]
Do we even need the condition that checks PlayerID?
Maybe @LordHansCapon can explain what it does.

**Comment by LordHansCapon** [comment]
@LeSpec The function I hooked has the current player invoking this function in the stack at `[esp+04]` (`esp+08` because I pushed ebx before it). I thought making sure that this is the currently controlled player will fix the issue in MP aswell. I guess not.

**Pull request merged**